src/schematools/validation.py

```diff
@@ -353,3 +390,58 @@ def _relation_auth(dataset: DatasetSchema) -> Iterator[str]:
                     scopes |= rel_table.get_field_by_id(f).auth
                 scopes.remove("OPENBAAR")  # Not very interesting.
                 yield f"{table.id}.{field.id} requires scopes {sorted(scopes)}"
+
+
+@_register_validator("reasons non public exists")
+def _reasons_non_public_exists(dataset: DatasetSchema) -> Iterator[str]:
+    """A ReasonsNonPublic field should be present on the highest non-public scope.
+
+    For non-public fields in a public table, reasonsNonPublic can be set on table
+    level. This is less verbose when all the non-public fields have the same
+    reason for being non-public.
+
+    """
+    if dataset.auth == {"OPENBAAR"}:
+        for table in dataset.tables:
+            if table.data.get("reasonsNonPublic") is None:
+                if table.auth == {"OPENBAAR"}:
+                    for field in table.get_fields():
+                        if (
+                            field.auth != {"OPENBAAR"}
+                            and field.data.get("reasonsNonPublic") is None
+                        ):
+                            yield (
+                                f"Non-public field {field.id} or it's parent table "
+                                "should have a 'reasonsNonPublic' property."
+                            )
+                else:
+                    yield (
+                        f"Non-public table {table.id} should have a 'reasonsNonPublic' property."
+                    )
+    elif dataset.data.get("reasonsNonPublic") is None:
+        yield (f"Non-public dataset {dataset.id} should have a 'reasonsNonPublic' property.")
+
```

Comment:
voor leesbaarheid zou ik de if/else logica omdraaien. de `if not openbaar: return` dan is er ook minder nesting nodig